plasma/limbo-priority-mod.md

```diff
@@ -0,0 +1,58 @@
+
+The effect of this design is that an in-flight transaction that may or may not be included in the Plasma chain (a transaction "in limbo") can now be guaranteed to be included. Furthermore, if all of a transactions inputs were created at a time when the chain was valid, then the transaction will be successfully processed before any invalid transactions are processed.
+
+## Construction
+
+A user may attempt to set the priority of a transaction by submitting a signed transaction and a block number to the root chain, subject to a challenge period. This action is bonded, and the priority of deposit ("out of nowhere") transactions cannot be modified. If no successful challenge occurs during the challenge period, then the transaction will have a priority as if it were the added to the end of the given block. 
+
+A priority modification can be challenged in four ways (higher priority = would be exited sooner, lower priority = would be exited later):
+
+1. Someone shows that the specified transaction was included in a block, and that its original priority is higher than the new priority. This prevents someone from decreasing a transaction's priority. 
+2. Someone shows that any of the transaction's inputs would have a lower priority than the transaction. This prevents someone from moving a transaction "earlier" than any of the transaction's outputs.
+3. Someone requests that someone prove some input to the transaction has a higher priority than new priority. This is necessary if blocks are being withheld and a challenge of type (2) is not directly possible.
+4. Someone shows that there exists another spend from any of the transaction's inputs. This spend must not be the same as the original transaction.
+
+A challenge of type (1) or (2) blocks the modification and slashes the submitter's bond. 
+
+A challenge of type (3) or (4) must be bonded, and are subject to a response.
+
+### Type (3) Response
+
+Anyone may respond to a type (3) challenge by revealing that the requested input has been included in the chain such that it has a higher priority than the new priority. A valid response would slash the challenger's bond. A failure to respond would slash the original bond and block the modification.
+
+### Type (4) Response
+
+Anyone may respond to a type (4) challenge by revealing another included transaction that spends from the original transaction's inputs. This revealed transaction may or may not be the same as the original transaction. A successful response slashes the challenger's bond. A failure to respond slashes the original bond, blocks the modification, and burns any inputs that have now been provably double spent. This is accomplished by adding the input to the priority queue and burning it when processing instead of sending it to its owner.
```

Comment:
The scenario I came up with is following.

a. [Alice, Mallory] -> [Bob] is signed and sent to operator. Operator sees large input from Alice, produces fraudulent block and withholds it.
b. Alice submits priority modification to root contract.
c. Mallory creates a double-spend for his input and challenges (type 4) modification.
d. Alice creates signs (Alice) -> (Bob) tx and submits its priority modification to root chain.
e. Alice uses modification (d) as response with **included** transaction.

Reply:
(c) must be included in the chain - so Mallory must cooperate with the operator. 

(d) doesn't count unless (Alice) -> (Bob) is already included (either in chain or via successfully priority modification).

Reply:
> (c) must be included in the chain - so Mallory must cooperate with the operator.

So type 4 challenge tx must be included in the chain. This is not clear from the text (word included is used in some places and not used in other places).

> (d) doesn't count unless (Alice) -> (Bob) is **already** included (either in chain or via successfully priority modification).

Can you elaborate on timing here?

Reply:
> So type 4 challenge tx must be included in the chain. 

Yep, sorry will make more clear.

> Can you elaborate on timing here?

So either it's included in the child chain at a position earlier than (Mallory) -> (X), or the modification of (Alice) -> (Bob) has completed (therefore "including" it in the child chain), but this won't happen before the modification of (Alice, Mallory) -> (Bob) because it was submitted first.

Reply:
Is Alice out of luck? Practically that would mean that we do not do gas optimizations with 3rd parties. And I think that atomic swaps for trades are affected.